### DISCUSSION TITLE
[Enhancement] Use gesture events instead of GestureRecognizers

### DISCUSSION THREAD

**Discussion by MichaelRumpler** [dotnet/maui#107] [General]
## Summary

When Xamarin.Forms was made, the developers had only iOS and Android on their minds. They also copied parts of those APIs which used GestureRecognizers / detectors. Objective-C and Java don't have events, so the architects of those APIs had to think of something else. But C# had events from the very start and these are much easier for being notified when something happens. Events have been used by .NET in every framework MS wrote.

## API Changes

Remove `X.F.View.GestureRecognizers` and add events and commands instead which handle all the gestures.

In [MR.Gestures](https://www.mrgestures.com/) I currently have `Down`, `Up`, `Tapping`, `Tapped`, `DoubleTapped`, `LongPressing`, `LongPressed`, `Panning`, `Panned`, `Swiped`, `Pinching`, `Pinched`, `Rotating` and `Rotated`. As you are advertising MAUI to be for desktop development (and mobile OSes also get better support for bluetooth mice), `PointerEntered`, `PointerMoved` and `PointerExited` also make sense.

When I made MR.Gestures I didn't think about nested gesture handling controls enough. I should've also added Preview- events. As I didn't have them, handling nested controls was always a PITA. If you also add those Preview- events and provide a method to handle the gesture in either the preview or the normal event, lives of your users would be easier.

I'm also willing to contribute my code of MR.Gestures. I didn't look at your gesture recognizers for a long time, but as people still buy MR.Gestures, it must still be better.

## Links

There are plenty of examples how MR.Gestures work on

- [https://www.mrgestures.com/](https://www.mrgestures.com/)
- [https://github.com/MichaelRumpler/GestureSample](https://github.com/MichaelRumpler/GestureSample)

This would be the same in MAUI.

**Comment by PureWeen** [comment]
We haven't had a chance to dissect how exactly we're going to manifest gestures but we will most likely keep the GestureRecognizers for compatibility reasons. 

> I'm also willing to contribute my code of MR.Gestures. I didn't look at your gesture recognizers for a long time, but as people still buy MR.Gestures, it must still be better.

Thank you for this!!! Once we get to gestures we'll start evaluating all of this closer

**Reply by roubachof** [reply]
I can only vouch for the quality of Mr Gestures. I use it in every app I made, it will work when XF gestures no longer works. And it has very nice fine grained control.

**Reply by roubachof** [reply]
Maybe we could have full MR.Gestures in https://github.com/xamarin/XamarinCommunityToolkit ?
And @MichaelRumpler could still help in maui improving quality of gestures ?

**Reply by WashingtonARamos** [reply]
Isn't this a great time to ditch the old (even if it means breaking compatibility which will happen anyway as there'll be needed project migrations) and embrace something new that makes much more sense in the C# context?
